### ISSUE TITLE
[Suggestion] Run database queries asynchronously

### ISSUE THREAD

**Issue by irenicus30** [otland/forgottenserver#2605]
Hi, I just saw this line of code [https://github.com/otland/forgottenserver/blob/master/src/database.cpp#L110](url) . But this is not even that important becouse this line gets executed in case of error. What looks more important to me is that almost all database queries (at least connected to player login) take place is dispatcher thread, which is main game logic thread. Those database queries take place in blocking manner (or am I wrong? I'm quite new to forgottenserver). So do you guys think it's possible to change that? Are there any important reasons why not implement database queries asynchronously, or at least in separate thread (except complexity of such solution)? Did somebody have issues related to those blocking database queries? (for example when database is on another machine)

**Comment by djarek** [comment]
The most important reason why queries are not invoked asynchronously is that the server was never designed to support that. While loading a player could be done asynchronously, saving is a little harder because you have to make sure the saved state is consistent across all the queries performed (e.g. to avoid allowing item duplication by relogging before a save completes.

TL;DR: it's *a lot* of work.
